File: 38.CountAndSay/Count.java
public class Count {
    public String countAndSay(int n) {
        if (n < 1) {
            return "";
        }
        String curr = "1";
        for (int i = 0; i < n - 1; i++) {
            curr = read(curr);
        }

        return curr;
    }

    public String read(String str) {
        if (str == null || str.isEmpty()) {
            return "";
        }
        char digit = str.charAt(0);
        int count = 1;
        StringBuilder res = new StringBuilder();
        for (int i = 1; i < str.length(); i++) {
            char curr = str.charAt(i);
            if (curr == digit) {
                count++;
            }
            else {
                res.append(count).append(digit);
                digit = curr;
                count = 1;
            }
        }
        res.append(count).append(digit);
        return res.toString();
    }

    public static void main(String[] args) {
        System.out.println(new Count().countAndSay(4));
    }
}

File: 38.CountAndSay/count.cc
#include <iostream>
#include <string>

using namespace std;

class Solution {
public:
    string countAndSay(int n) {
        if (n < 1) {
            return "";
        }

        string curr = "1";
        for (int i = 0; i < n - 1; i++) {
            curr = read(curr);
        }

        return curr;
    }

    string read(string str) {
        if (str.size() == 0) {
            return "";
        }

        char digit = str[0];
        int count = 1;
        string res = "";
        for (int i = 1; i < int(str.size()); i++) {
            if (str[i] == digit) {
                count++;
            }
            else {
                res.push_back(char(count+'0'));
                res.push_back(digit);
                digit = str[i];
                count = 1;
            }
        }
        res.push_back(char(count+'0'));
        res.push_back(digit);

        return res;
    }
};

int main(int argc, char const *argv[]) {
    Solution s = Solution();
    cout << s.countAndSay(4) << endl;
    return 0;
}
